### PR TITLE
Feature/cloud trails

### LINKED ISSUE #237: Implement CloudTrail Configuration
Some Clients have requested monitoring / alerting on specific activities done with the account, for example, alerting if an S3 bucket is changed to public.

The AWS recommendation is a CloudTrail pipeline.

I did not see how to configure CloudTrail in Gyro.
https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-user-guide.html

### PR DESCRIPTION
**Pull request by harjain99**
Fixes #237

### CODE REVIEW

File: src/main/java/gyro/aws/cloudtrail/CloudTrailDataResource.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    public void setType(String type) {
+        this.type = type;
+    }
+
+    /**
+     * The list of Amazon Resource Names (ARNs) of the resources in which the data events should be logged. (Required)
+     */
+    @Updatable
+    @Required
+    public List<String> getValues() {
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/cloudtrail/CloudTrailDataResource.java:52]: As both these are part of the primary key, they can essentially never be updated, as a change will trigger a delete an create

File: src/main/java/gyro/aws/cloudtrail/TrailResource.java [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+        if (changedFieldNames.contains("log-group-arn")) {
+            builder = builder.cloudWatchLogsLogGroupArn(getLogGroupArn());
+        }
+
+        if (changedFieldNames.contains("is-organization-trail")) {
+            builder = builder.isOrganizationTrail(getIsOrganizationTrail());
+        }
+
+        if (changedFieldNames.contains("logs-role")) {
+            builder = builder.cloudWatchLogsRoleArn(getLogsRole().getArn());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/cloudtrail/TrailResource.java:411]: Potential NPE!

File: src/main/java/gyro/aws/cloudtrail/TrailResource.java [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+        }
+
+        if (changedFieldNames.contains("key")) {
+            builder = builder.kmsKeyId(getKey().getArn());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/cloudtrail/TrailResource.java:415]: Potential NPE!

File: src/main/java/gyro/aws/cloudtrail/TrailResource.java [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+        CloudTrailClient client = createClient(CloudTrailClient.class);
+
+        UpdateTrailRequest.Builder builder = UpdateTrailRequest.builder().name(getName());
+
+        if (changedFieldNames.contains("bucket")) {
+            builder = builder.s3BucketName(getBucket().getName());
+        }
+
+        if (changedFieldNames.contains("bucket-key-prefix")) {
+            builder = builder.s3KeyPrefix(getBucketKeyPrefix());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/cloudtrail/TrailResource.java:383]: Potential NPE!

File: src/main/java/gyro/aws/cloudtrail/CloudTrailEventSelector.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        setDataResource(model.dataResources().stream().map(r -> {
+            CloudTrailDataResource cloudTrailDataResource = newSubresource(CloudTrailDataResource.class);
+            cloudTrailDataResource.copyFrom(r);
+            return cloudTrailDataResource;
+        }).collect(Collectors.toList()));
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format("Data Resource: %s, Excluded Sources: %s", getDataResource().toString(), getManagementEventSourcesToExclude().toString());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/cloudtrail/CloudTrailEventSelector.java:108]: Both of these are lists, so how does two Cloudtrail event selector actually differ ?

File: src/main/java/gyro/aws/cloudtrail/CloudTrailDataResource.java (hunk 1 above)
Comment (harjain99) [reply src/main/java/gyro/aws/cloudtrail/CloudTrailDataResource.java:52]: That means that the Updatable annotation isn't needed here, right?
I'll remove it.

File: examples/cloudtrail/trail.gyro [hunk 6]
```diff
@@ -0,0 +1,8 @@
+aws::trail example-trail
+    name: "example-trail"
+    bucket: "example-bucket"
+
+    tags: {
+        "example-key": "example-value"
+    }
+end
```
Comment (deepanjan90) [inline examples/cloudtrail/trail.gyro:8]: Should contain the example for event selector and insight selector

File: src/main/java/gyro/aws/cloudtrail/TrailResource.java [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+ * .. code-block:: gyro
+ *
+ *     aws::trail example-trail
+ *         name: "example-trail"
+ *         bucket: "example-bucket"
+ *
+ *         tags: {
+ *             "example-key": "example-value"
+ *         }
+ *     end
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/cloudtrail/TrailResource.java:64]: Should contain the example for event selector and insight selector

File: src/main/java/gyro/aws/cloudtrail/TrailFinder.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * Example
+ * -------
+ *
+ * .. code-block:: gyro
+ *
+ *    trail: $(external-query aws::trail { name: "example-trail" })
+ */
+@Type("trail")
+public class TrailFinder extends AwsFinder<CloudTrailClient, Trail, TrailResource> {
```
Comment (beetlebugorg) [inline src/main/java/gyro/aws/cloudtrail/TrailFinder.java:42]: Rename this to `CloudTrailFinder` and `@Type("cloudtrail").

File: src/main/java/gyro/aws/cloudtrail/TrailResource.java [hunk 9]
```diff
@@ -0,0 +1,587 @@
...
+ *             insight-type: "ApiCallRateInsight"
+ *         end
+ *
+ *         tags: {
+ *             "example-key": "example-value"
+ *         }
+ *     end
+ */
+@Type("trail")
+public class TrailResource extends AwsResource implements Copyable<Trail> {
```
Comment (beetlebugorg) [inline src/main/java/gyro/aws/cloudtrail/TrailResource.java:87]: Rename this to `CloudTrailResource` and `@Type("cloudtrail").

**Pull request merged**